text/3773-must-use-output.md

```diff
@@ -0,0 +1,124 @@
+
+```rust
+impl<T> Vec<T> {
+    pub fn push(#[must_use_output] &mut self, item: T) { /* ... */ }
+}
+```
+
+This is conceptually similar to `#[must_use]` on the function, but for an output parameter rather than the return value. Like `#[must_use]`, it helps you (or downstream consumers) to catch likely mistakes.
+
+A common case when this is useful is when the function has no side effects other than mutation through the passed reference and allocation. A typical example is methods modifying collections, builders or similar objects.
+
+Note that this is currently subject to some limitations: if the reference was passed in as an argument, returned from another function or obtained from a pointer the warning will not trigger. This may change in the future and it may start producing more warnings in some cases.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Whenever a function parameter is marked with `#[must_use_output]`, and the caller does not later access the value, a warning is emitted similar to using the `+=` operator on integers and later not using the value.
+
+It is an error to put this attribute on parameters with types other than reference types. Raw pointers, smart pointers, and generics that hide the reference, are forbidden, mainly for simplicity of initial implementation. They can be implemented and allowed later, if it's even possible at all.
+
+The compiler makes no distinction whether the reference is unique or shared because writes can happen through both; there doesn't seem to be a reason to forbid `Freeze` references and there is at least one case when even a `Freeze` reference is useful.
+
+The standard library functions get annotated with this attribute as well, including but not limited to:
+
+- Modifications of collections that don't return a value
+- `core::mem::swap`
+- `Clone::clone`
+- `*_assign` methods in `core::ops` traits
+- `OpenOptions` and similar builders
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- Adds another attribute to the language, complicating it
+- In some cases analysis will fail and perhaps people will have false sene of security
+- People could misuse it to attempt to enforce soundness of `unsafe` code (which it cannot do)
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+- We could also simply not do this but the potential mistakes it catches are real.
+- The name could be something different and preferably shorter. The name used here was suggested by Josh Triplett and is pretty readable and it can be used before stabilization in the absence of better one. A pretty short and clear `#[must_read]` was also suggested by Kevin Reid.
+- We could write `#[must_use]` on a reference parameter instead. The downside would be that this could be mistaken for saying that the *callee* must use the parameter, rather than the *caller*.
```

Comment:
This doesn't seem like much of a problem to me.  We already lint on unused function parameters (except for `self`, unfortunately, due to not having a good way to suppress it), so if I saw `#[must_use]` on a `&mut` parameter, I'd expect this would refer to an obligation on the caller, especially as that's the existing connotation of `#[must_use]`.

Barring better proposals (and hopefully more concise ones than `#[must_use_output]`), my own preference is to just use `#[must_use]` for this.

Reply:
Yeah, I also feel like it wouldn't be that big deal and people would figure it out quickly. But at the same time I'm not bothered by the name too much and I respect whatever decision will be made by the appropriate team.

Reply:
I also prefer `#[must_use]`. When applied to a function, it also means "the output at the call site must be used", and people don't have a problem understanding it.

Reply:
Maybe the attribute could be named differently, something along the lines of the function having side effect that needs to be observed `#[must_use(side_effect)]` or `#[is_a_setter]`?

Reply:
We have a couple of parallel discussions occurring about the name, one here and one at https://github.com/rust-lang/rfcs/pull/3773#discussion_r1974960959 .

From that discussion: `caller_must_use` seems like it removes the ambiguity.